libsolidity/formal/CHC.cpp

```diff
@@ -1172,32 +1166,15 @@ void CHC::resetSourceAnalysis()
 	m_blockCounter = 0;
 
 	// At this point every enabled solver is available.
-	// If more than one Horn solver is selected we go with z3.
-	// We still need the ifdef because of Z3CHCInterface.
-	if (m_settings.solvers.z3)
-	{
-#ifdef HAVE_Z3
-		// z3::fixedpoint does not have a reset mechanism, so we need to create another.
-		m_interface = std::make_unique<Z3CHCInterface>(m_settings.timeout);
-		auto z3Interface = dynamic_cast<Z3CHCInterface const*>(m_interface.get());
-		solAssert(z3Interface, "");
-		m_context.setSolver(z3Interface->z3Interface());
-#else
-		solAssert(false);
-#endif
-	}
-	if (!m_settings.solvers.z3)
-	{
-		solAssert(m_settings.solvers.smtlib2 || m_settings.solvers.eld);
+	solAssert(m_settings.solvers.eld || m_settings.solvers.z3);
 
-		if (!m_interface)
-			m_interface = std::make_unique<CHCSmtLib2Interface>(m_smtlib2Responses, m_smtCallback, m_settings.solvers, m_settings.timeout);
+	if (!m_interface)
+		m_interface = std::make_unique<CHCSmtLib2Interface>(m_smtlib2Responses, m_smtCallback, m_settings.solvers, m_settings.timeout);
 
-		auto smtlib2Interface = dynamic_cast<CHCSmtLib2Interface*>(m_interface.get());
-		solAssert(smtlib2Interface, "");
-		smtlib2Interface->reset();
-		m_context.setSolver(smtlib2Interface->smtlib2Interface());
-	}
+	auto smtlib2Interface = dynamic_cast<CHCSmtLib2Interface*>(m_interface.get());
```

Comment:
Do we plan to use any  CHC solver that does not support SMTLib2 interface?
If not, maybe we could change `m_interface` type to `CHCSmtLib2Interface`. That would simplify things a bit.